src/client/ua_client.c

```diff
@@ -673,29 +725,71 @@ UA_Client_backgroundConnectivity(UA_Client *client) {
         client->pendingConnectivityCheck = true;
 }
 
-static void
-clientExecuteRepeatedCallback(void *executionApplication, UA_ApplicationCallback cb,
-                              void *callbackApplication, void *data) {
-    cb(callbackApplication, data);
+UA_StatusCode
+UA_Client_setupEventLoop(UA_Client *client) {
+
+    UA_BasicClientConnectionContext *ctx = (UA_BasicClientConnectionContext*) UA_malloc(sizeof(UA_BasicClientConnectionContext));
+    UA_CHECK_MEM(ctx, return UA_STATUSCODE_BADOUTOFMEMORY);
+    memset(ctx, 0, sizeof(UA_BasicClientConnectionContext));
+    client->connection.state = UA_CONNECTIONSTATE_CLOSED;
+    ctx->client = client;
+    ctx->isInitial = true;
+
+    UA_ConnectionManager *cm = UA_ConnectionManager_TCP_new(UA_STRING("tcpCM"));
+    UA_CHECK_MEM(cm, return UA_STATUSCODE_BADOUTOFMEMORY);
+
+    cm->shutdownCallback = UA_Client_shutdownCallback;
```

Comment:
connection managers also have shutdown callbacks now... because otherwise specific connection cleanups were not possible.